### PR TITLE
state update in each iteration for perceptron

### PR DESCRIPTION
**Pull request by shubham808**
We want to update the weights vector in each iteration so that we remain consistent during a paused state

### CODE REVIEW

File: src/shogun/classifier/Perceptron.cpp [hunk 1]
```diff
@@ -35,7 +35,9 @@ void CPerceptron::init()
 	max_iter = 1000;
 	learn_rate = 0.1;
 	m_initialize_hyperplane = true;
-	SG_ADD(&max_iter, "initialize_hyperplane", "Whether to initialize hyperplane.", MS_AVAILABLE);
```
Comment (karlnapf) [inline src/shogun/classifier/Perceptron.cpp]: why the change? just by editor accident?

File: src/shogun/classifier/Perceptron.cpp [hunk 2]
```diff
@@ -65,15 +67,19 @@ bool CPerceptron::train_machine(CFeatures* data)
 	ASSERT(num_vec==train_labels.vlen)
 	SGVector<float64_t> output(num_vec);
 
-	SGVector<float64_t> w = get_w();
+	auto w = SGVector<float64_t>(num_feat);
+	set_w(w);
```
Comment (karlnapf) [inline src/shogun/classifier/Perceptron.cpp]: this is bad.
You over-write the previous w in any case, which kills the feature of users being able to initialise themselves.
The setting should go into the else branch of the if statement.

File: src/shogun/classifier/Perceptron.cpp [hunk 3]
```diff
@@ -65,15 +67,19 @@ bool CPerceptron::train_machine(CFeatures* data)
...
 	if (m_initialize_hyperplane)
 	{
-		w = SGVector<float64_t>(num_feat);
+		w.set_const(0);
```
Comment (karlnapf) [inline src/shogun/classifier/Perceptron.cpp]: set const to 1.0 / num_feat and them remove the `add` below

File: src/shogun/classifier/Perceptron.cpp (hunk 1 above)
Comment (shubham808) [reply src/shogun/classifier/Perceptron.cpp]: see that its &max_iter and not &m_initialize_hyperplane

Comment (karlnapf) [reply src/shogun/classifier/Perceptron.cpp]: ah yes, correcting copy paste errors. nice!

File: src/shogun/classifier/Perceptron.cpp [hunk 4]
```diff
@@ -65,15 +67,18 @@ bool CPerceptron::train_machine(CFeatures* data)
 	ASSERT(num_vec==train_labels.vlen)
 	SGVector<float64_t> output(num_vec);
 
-	SGVector<float64_t> w = get_w();
+	auto w = SGVector<float64_t>(num_feat);
```
Comment (karlnapf) [inline src/shogun/classifier/Perceptron.cpp]: The allocation should be moved into the if

Comment (karlnapf) [reply src/shogun/classifier/Perceptron.cpp]: i.e. just define it here

Comment (karlnapf) [reply src/shogun/classifier/Perceptron.cpp]: actually, why dont you just do m_w.resize(num_feats)?

Comment (shubham808) [reply src/shogun/classifier/Perceptron.cpp]: m_w is private so we will need to use get_w() everywhere instead... this causes type errors with v.add(gradient, w)

Comment (karlnapf) [reply src/shogun/classifier/Perceptron.cpp]: ok cool. But still you can move the allocation inside the if, you dont want to allocate memory if the user has provided intitialization

File: src/shogun/classifier/Perceptron.cpp [hunk 5]
```diff
@@ -65,15 +67,18 @@ bool CPerceptron::train_machine(CFeatures* data)
...
 	if (m_initialize_hyperplane)
 	{
-		w = SGVector<float64_t>(num_feat);
```
Comment (karlnapf) [inline src/shogun/classifier/Perceptron.cpp]: keep this line here, and change the above to a definition

File: tests/unit/classifier/Perceptron_unittest.cc [hunk 6]
```diff
@@ -59,3 +59,25 @@ TEST(Perceptron, train)
...
+
+	auto perceptron = some<CPerceptron>(features, labels);
+	perceptron->train();
+
+	auto weights = perceptron->get_w();
+
+	auto perceptron_initialized = some<CPerceptron>(features, labels);
+	perceptron_initialized->set_initialize_hyperplane(false);
+	perceptron_initialized->set_w(weights);
+	perceptron_initialized->set_max_iter(1);
```
Comment (karlnapf) [inline tests/unit/classifier/Perceptron_unittest.cc]: yep this seems to make sense

File: tests/unit/classifier/Perceptron_unittest.cc [hunk 7]
```diff
@@ -59,3 +59,25 @@ TEST(Perceptron, train)
...
+
+	perceptron_initialized->train();
+	EXPECT_EQ(perceptron_initialized->get_w(), weights);
```
Comment (karlnapf) [inline tests/unit/classifier/Perceptron_unittest.cc]: this checks for ==, can you make it equals?

Comment (karlnapf) [reply tests/unit/classifier/Perceptron_unittest.cc]: `EXPECT_TRUE(perceptron_initialized->get_w().equals(weights))`
== is pointer equality (should fail)

Comment (shubham808) [reply tests/unit/classifier/Perceptron_unittest.cc]: i see

**Comment by karlnapf** [comment]
Waiting for CI and then good to merge

**Comment by shubham808** [comment]
any changes here ? :)

**Comment by geektoni** [comment]
It seems okay from my side. @karlnapf?
